v2/internal/runner/proxy.go

```diff
@@ -0,0 +1,145 @@
+			scanner := bufio.NewScanner(file)
+			for scanner.Scan() {
+				proxy := scanner.Text()
+				if strings.TrimSpace(proxy) == "" {
+					continue
+				}
+				if proxyURL, err := validateProxyURL(proxy); err != nil {
+					return err
+				} else {
+					proxyURLList = append(proxyURLList, proxyURL)
+				}
+			}
+		} else {
+			return fmt.Errorf("invalid proxy file or URL provided for %s", p)
+		}
+	}
+	return processProxyList(options)
+}
+
+func processProxyList(options *types.Options) error {
+	if len(proxyURLList) == 0 {
+		return fmt.Errorf("could not find any valid proxy")
+	} else {
+		done := make(chan bool)
+		exitCounter := make(chan bool)
+		counter := 0
+		for _, url := range proxyURLList {
+			go runProxyConnectivity(url, options, done, exitCounter)
+		}
+		for {
+			select {
+			case <-done:
+				{
+					close(done)
+					return nil
+				}
+			case <-exitCounter:
+				{
+					if counter += 1; counter == len(proxyURLList) {
+						return errors.New("no reachable proxy found")
+					}
+				}
+			}
+		}
+	}
+}
+
+func runProxyConnectivity(proxyURL url.URL, options *types.Options, done chan bool, exitCounter chan bool) {
+	if err := testProxyConnection(proxyURL, options.Timeout); err == nil {
+		// if options.ProxyURL == "" && options.ProxySocksURL == "" {
+		// 	if valid := assignProxyURL(proxyURL, options); valid {
+		// 		done <- true
+		// 	}
+		// }
+		if os.Getenv(types.HTTP_PROXY_ENV) == "" {
+			if valid := assignProxyURL(proxyURL, options); valid {
+				done <- true
+			}
+		}
+	}
+	exitCounter <- true
+}
+
+func testProxyConnection(proxyURL url.URL, timeoutDelay int) error {
+	timeout := time.Duration(timeoutDelay) * time.Second
+	_, err := net.DialTimeout("tcp", fmt.Sprintf("%s:%s", proxyURL.Hostname(), proxyURL.Port()), timeout)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func assignProxyURL(proxyURL url.URL, options *types.Options) bool {
+	var isValid bool = true
+	if proxyURL.Scheme == types.HTTP || proxyURL.Scheme == types.HTTPS {
+		os.Setenv(types.HTTP_PROXY_ENV, proxyURL.String())
+		// options.ProxyURL = proxyURL.String()
+		// options.ProxySocksURL = ""
+		gologger.Verbose().Msgf("Using %s as proxy server", proxyURL.String())
+	} else if proxyURL.Scheme == types.SOCKS5 {
+		os.Setenv(types.HTTP_PROXY_ENV, proxyURL.String())
+		// options.ProxyURL = ""
+		// options.ProxySocksURL = proxyURL.String()
+		gologger.Verbose().Msgf("Using %s as socket proxy server", proxyURL.String())
+	} else {
+		isValid = false
+	}
+	return isValid
+}
+
+func validateProxyURL(proxy string) (url.URL, error) {
+	if url, err := url.Parse(proxy); err == nil && isSupportedProtocol(url.Scheme, false) {
+		return *url, nil
+	}
+	return url.URL{}, errors.New("invalid proxy format (It should be http[s]/socks5://[username:password@]host:port)")
+}
+
+//isSupportedProtocol checks given protocols are supported
+func isSupportedProtocol(value string, prefixCheck bool) bool {
+	if prefixCheck {
+		value = strings.ToLower(value)
```

Comment:
The `url.Scheme` should already contain the value lower cased. I'd change the method parameter to `url` and then you can remove the `ToLower` part

Reply:
Done